### PR TITLE
Add `certgen` subcommand with output to PEM format.

### PR DESCRIPTION
**Pull request by youngnick**
Updates #1215 

Signed-off-by: Nick Young <ynick@vmware.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -133,6 +133,7 @@ cleancerts:
 
 certs/CAkey.pem:
 	@echo No CA keypair present, generating
+	@mkdir -p certs
```
Comment (davecheney) [inline Makefile]: makefile pro tip
```
certs/CAkey.pem: | certs
   ...

certs:
     @mkdir -p certs
```
this makes the _presence_ of `certs/` a prereq for certs/CAKey.pem, but make won't be picky about timestamps

File: cmd/contour/certgen.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Copyright © 2018 Heptio
```
Comment (davecheney) [inline cmd/contour/certgen.go]: 2019

File: cmd/contour/contour.go [hunk 3]
```diff
@@ -96,6 +99,16 @@ func main() {
 	switch kingpin.MustParse(app.Parse(args)) {
 	case bootstrap.FullCommand():
 		doBootstrap(bootstrapCtx)
+	case certgenApp.FullCommand():
+		generatedCerts, err := certgen.GenerateCerts(certgenConfig)
+		if err != nil {
```
Comment (davecheney) [inline cmd/contour/contour.go]: check(err)

File: cmd/contour/contour.go [hunk 4]
```diff
@@ -96,6 +99,16 @@ func main() {
...
+			// TODO(youngnick) placeholder, fix this.
+			log.Fatal(err)
+		}
+		err = certgen.OutputCerts(certgenConfig, generatedCerts)
```
Comment (davecheney) [inline cmd/contour/contour.go]: same

File: internal/certgen/certgen.go [hunk 5]
```diff
@@ -0,0 +1,331 @@
+// Copyright © 2018 Heptio
```
Comment (davecheney) [inline internal/certgen/certgen.go]: 2019

File: internal/certgen/certgen.go [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+	"io"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const keySize = 2048
```
Comment (davecheney) [inline internal/certgen/certgen.go]: please add a comment to explain why 2048 is the magic number

File: internal/certgen/certgen.go [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+	log := certgenConfig.FieldLogger
+
+	if certgenConfig.OutputPEM {
+		// TODO(youngnick): Should we sanitize this value?
+		err := os.MkdirAll(certgenConfig.OutputDir, 0755)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("Outputting certs to PEM files in %s/", certgenConfig.OutputDir)
```
Comment (davecheney) [inline internal/certgen/certgen.go]: I'd drop the logger here and just use straigth fmt.Printf for two reasons

1. log goes to stderr
2. nobody needs timestamps in their "here's your files" line

File: internal/certgen/certgen_test.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
+// Copyright © 2018 Heptio
```
Comment (davecheney) [inline internal/certgen/certgen_test.go]: 2019

File: internal/certgen/certgen_test.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func verifyCert(certPEM []byte, roots *x509.CertPool, dnsname string) error {
+	block, _ := pem.Decode(certPEM)
+	if block == nil {
+		return fmt.Errorf("Failed to decode %s certificate from PEM form", dnsname)
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
```
Comment (davecheney) [inline internal/certgen/certgen_test.go]: why not return err here, it probably says more than the message on line 77

File: internal/certgen/certgen_test.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		return errors.New("Failed to parse an x509 certificate out of the Contour cert")
+	}
+
+	opts := x509.VerifyOptions{
+		DNSName: dnsname,
+		Roots:   roots,
+	}
+	if _, err = cert.Verify(opts); err != nil {
```
Comment (davecheney) [inline internal/certgen/certgen_test.go]: why not just

return cert.Verify(opts)

if cert.Verify works then it'll return nil, like line 88. If it fails, the err message probably already says "verification failed"

File: cmd/contour/certgen.go (hunk 2 above)
Comment (youngnick) [reply cmd/contour/certgen.go]: 🤦‍♂

File: internal/certgen/certgen_test.go (hunk 10 above)
Comment (youngnick) [reply internal/certgen/certgen_test.go]: Verify returns two values, the first is the complete certificate chain (which we don't care about here), so I'll need to leave this unless there's way to throw away the first value that's shorter than this.

**Pull request merged**